---
title: "SCOP_2021_0177: Transcriptomic analysis of DNA methylation edited mouse neural stem cells"
author: 
- "Analysis by Thomas Gade Koefoed"
- "Staff Scientist at the Single-Cell Omics Platform"
- "thomas.koefoed@sund.ku.dk"
date: "`r format(Sys.time(), '%d %B, %Y')`"
knit: (function (...) {
        rmarkdown::render(
          basename(rstudioapi::getSourceEditorContext()$path), 
          output_file = paste(
            format(Sys.time(),'%Y_%m_%d'),
            tools::file_path_sans_ext(basename(rstudioapi::getSourceEditorContext()$path)),
            sep="_"), 
          envir = new.env())})
output:
   html_document:
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: hide
    df_print: paged
editor_options:
  chunk_output_type: inline
---

# Code blocks: Initial data processing {.tabset}
All code used for the analyses can be found throughout this report by expanding the code-blocks in the .html-file on the right-hand side within the appropriate tabs, or through the associated .Rmd file, where it can also be run. You will have to install all the necessary packages yourself though, which can be non-trivial.

## Libraries
```{r, warning=FALSE, message=FALSE, results="hide"}
suppressPackageStartupMessages({
  library('biomaRt') # downloading gene data
  library("DT") # for pretty printing of paged tables
  library("limma") # RNA-seq analysis
  library("edgeR") # RNA-seq analysis
  library("DESeq2") # RNA-seq analysis
  library("here") # Better structure and relative paths
  library("GO.db") # downloading gene data
  library("org.Mm.eg.db")
  library("tidyverse") # general data manipulation
  library("cbmr") # in-house convenience functions
  library("AnnotationDbi")
  library("ggrepel")
})

full_script_name <- rstudioapi::getSourceEditorContext()$path
script_name <- basename(full_script_name)
data_dir <- paste0(tools::file_path_sans_ext(full_script_name),"_data/")
dir.create(data_dir, showWarnings = F)
```


## Group method
```{r, message = FALSE, class.source = 'fold-show'}
# load data
g_counts_raw <- readRDS("../data-raw/salmon.merged.gene_counts.rds")
rna_counts_all <- g_counts_raw %>% assays() %>% .$counts

# fix sample IDS in count matrix
colnames(rna_counts_all) <- 
  sub(pattern = "X[0-9]{4}_", replacement = "", x = colnames(rna_counts_all)) %>% 
  sub(pattern = "_S[0-9]{1,2}", replacement = "", x=.)

# fix sample IDS in meta data sheet
meta_data_sheet <- readxl::read_excel("0177_metadata.xlsx") %>%
  transmute(`Sample.ID`=sub(pattern = ".....", replacement = "", x=`Sample ID (Library ID)`), 
            gRNA = factor(gsub(x = Condition1, pattern = "g|gRNA_", replacement = ""), levels=c("EV", "PTPRK", "PHOX2B")),
            Media = factor(gsub(x = Condition2, pattern = "_media", replacement = ""), levels=c("proliferation","differentiation"))) %>%
  as.data.frame()

#Create DGE list of relevant samples
rna_counts <- rna_counts_all[, as.character(c(1:18))]

y <- DGEList(rna_counts, 
             group = factor(paste0(meta_data_sheet$gRNA, meta_data_sheet$Media)),
             samples=meta_data_sheet)

design <-  model.matrix(~0 + group, data=y$samples)

idx_expressed <- filterByExpr(y, design = design)
y <- y[idx_expressed, ]
y <- calcNormFactors(y)

y <- estimateDisp(y, design = design)
fit <- glmQLFit(y, design = y$design, robust = TRUE)
group_fit <- fit

contrast_matrix <- makeContrasts(PHOX2B_vs_EV_proliferation=c(0,-1,0,1,0,0),
                                 PTPRK_vs_EV_proliferation=c(0,-1,0,0,0,1),
                                 PHOX2B_vs_EV_differentiation=c(-1,0,1,0,0,0),
                                 PTPRK_vs_EV_differentiation=c(-1,0,0,0,1,0),
  PHOX2B_vs_EV_all=c(-1,-1,1,1,0,0),
                                 PTPRK_vs_EV_all=c(-1,-1,0,0,1,1),
                                 
                                 
                                 levels=design)
                                 
all_genes_test_full_info <- list()
for (contrast in colnames(contrast_matrix)) {
  all_genes_test <- edgeR_tester(contrast_matrix = contrast_matrix, contrast=contrast, efit = fit)
  all_genes_test_full_info[[contrast]] <-  merge(x=all_genes_test, 
                                                 y=gene_info_list, 
                                                 by.x = "ENSEMBL", by.y="ensembl_gene_id",
                                                 all.x=T)
}
all_genes_test_full_info %>% lapply(function(x) {x %>% filter(FDR<0.05) %>% nrow()})

withr::with_par(list(mfrow = c(1,2)),
                {for (contrast in colnames(contrast_matrix)) {
                  hist(all_genes_test_full_info[[contrast]]$PValue,
                       main = contrast, xlab = "P-value") }}
)
```

## Batch effect
```{r, message = FALSE, class.source = 'fold-show'}
# load data
g_counts_raw <- readRDS("../data-raw/salmon.merged.gene_counts.rds")
rna_counts_all <- g_counts_raw %>% assays() %>% .$counts

# fix sample IDS in count matrix
colnames(rna_counts_all) <- 
  sub(pattern = "X[0-9]{4}_", replacement = "", x = colnames(rna_counts_all)) %>% 
  sub(pattern = "_S[0-9]{1,2}", replacement = "", x=.)

# fix sample IDS in meta data sheet
meta_data_sheet <- readxl::read_excel("0177_metadata.xlsx") %>%
  transmute(`Sample.ID`=sub(pattern = ".....", replacement = "", x=`Sample ID (Library ID)`), 
            gRNA = factor(gsub(x = Condition1, pattern = "g|gRNA_", replacement = ""), levels=c("EV", "PTPRK", "PHOX2B")),
            Media = factor(gsub(x = Condition2, pattern = "_media", replacement = ""), levels=c("proliferation","differentiation"))) %>%
  as.data.frame()

#Create DGE list of relevant samples
rna_counts <- rna_counts_all[, as.character(c(1:18))]

y <- DGEList(rna_counts, 
             group = factor(paste0(meta_data_sheet$gRNA)),
             samples=meta_data_sheet)

(design <-  model.matrix(~0 + group+Media, data=y$samples))

idx_expressed <- filterByExpr(y, design = design)
y <- y[idx_expressed, ]
y <- calcNormFactors(y)

y <- estimateDisp(y, design = design)
fit <- glmQLFit(y, design = y$design, robust = TRUE)
batch_fit <- fit

contrast_matrix <- makeContrasts(PTPRK_vs_EV_proliferation=groupPTPRK-groupEV,
                                 PHOX2B_vs_EV_proliferation=groupPHOX2B-groupEV,
                                 PTPRK_vs_EV_differentiation=groupPTPRK+Mediadifferentiation-(groupEV+Mediadifferentiation),
                                 PHOX2B_vs_EV_differentiation=groupPHOX2B+Mediadifferentiation-(groupEV+Mediadifferentiation),
                                 levels=design)

all_genes_test_full_info <- list()
for (contrast in colnames(contrast_matrix)) {
  all_genes_test <- edgeR_tester(contrast_matrix = contrast_matrix, contrast=contrast, efit = fit)
  all_genes_test_full_info[[contrast]] <-  merge(x=all_genes_test, 
                                                 y=gene_info_list, 
                                                 by.x = "ENSEMBL", by.y="ensembl_gene_id",
                                                 all.x=T)
}
all_genes_test_full_info %>% lapply(function(x) {x %>% filter(FDR<0.05) %>% nrow()})
withr::with_par(list(mfrow = c(1,2)),
                {for (contrast in colnames(contrast_matrix)) {
                  hist(all_genes_test_full_info[[contrast]]$PValue,
                       main = contrast, xlab = "P-value") }})
```

## Synergistic batch effect
```{r, message = FALSE, class.source = 'fold-show'}
# load data
g_counts_raw <- readRDS("../data-raw/salmon.merged.gene_counts.rds")
rna_counts_all <- g_counts_raw %>% assays() %>% .$counts

# fix sample IDS in count matrix
colnames(rna_counts_all) <- 
  sub(pattern = "X[0-9]{4}_", replacement = "", x = colnames(rna_counts_all)) %>% 
  sub(pattern = "_S[0-9]{1,2}", replacement = "", x=.)

# fix sample IDS in meta data sheet
meta_data_sheet <- readxl::read_excel("0177_metadata.xlsx") %>%
  transmute(`Sample.ID`=sub(pattern = ".....", replacement = "", x=`Sample ID (Library ID)`), 
            gRNA = factor(gsub(x = Condition1, pattern = "g|gRNA_", replacement = ""), levels=c("EV", "PTPRK", "PHOX2B")),
            Media = factor(gsub(x = Condition2, pattern = "_media", replacement = ""), levels=c("proliferation","differentiation"))) %>%
  as.data.frame()

#Create DGE list of relevant samples
rna_counts <- rna_counts_all[, as.character(c(1:18))]

y <- DGEList(rna_counts, 
             group = factor(paste0(meta_data_sheet$gRNA)),
             samples=meta_data_sheet)

(design <-  model.matrix(~0 + group*Media, data=y$samples))

colnames(design) <- gsub(colnames(design), pattern = ":", replacement = "")

idx_expressed <- filterByExpr(y, design = design)
y <- y[idx_expressed, ]
y <- calcNormFactors(y)

y <- estimateDisp(y, design = design)
fit <- glmQLFit(y, design = y$design, robust = TRUE)
synergistic_fit <- fit

contrast_matrix <- makeContrasts(PTPRK_vs_EV_proliferation=groupPTPRK-groupEV,
                                 PHOX2B_vs_EV_proliferation=groupPHOX2B-groupEV,
                                 PTPRK_vs_EV_differentiation=groupPTPRK+Mediadifferentiation+groupPTPRKMediadifferentiation-(groupEV+Mediadifferentiation),
                                 PHOX2B_vs_EV_differentiation=groupPHOX2B+Mediadifferentiation+groupPHOX2BMediadifferentiation-(groupEV+Mediadifferentiation),
                                 Mediadifferentiation,
                                 groupPTPRKMediadifferentiation,
                                 groupPHOX2BMediadifferentiation,
                                 
                                 levels=design)



all_genes_test_full_info <- list()
for (contrast in colnames(contrast_matrix)) {
  all_genes_test <- edgeR_tester(contrast_matrix = contrast_matrix, contrast=contrast, efit = fit)
  all_genes_test_full_info[[contrast]] <-  merge(x=all_genes_test, 
                                                 y=gene_info_list, 
                                                 by.x = "ENSEMBL", by.y="ensembl_gene_id",
                                                 all.x=T)
}
all_genes_test_full_info %>% lapply(function(x) {x %>% filter(FDR<0.05) %>% nrow()})
withr::with_par(list(mfrow = c(1,3)),
                {for (contrast in colnames(contrast_matrix)) {
                  hist(all_genes_test_full_info[[contrast]]$PValue,
                       main = contrast, xlab = "P-value") }})
```